tensorvis/tensorvisbase/include/inviwo/tensorvisbase/datastructures/tensorfield.h

```diff
@@ -0,0 +1,441 @@
+
+    template <typename T = size_t>
+    glm::vec<N, T> getBounds() const;
+
+    template <typename T = float>
+    glm::vec<N, T> getSpacing() const;
+
+    size_t getSize() const { return size_; }
+
+    matNb getBasisAndOffset() const;
+
+    std::shared_ptr<const std::vector<matN>> tensors() const;
+
+    void setMask(const std::vector<glm::uint8>& mask) { binaryMask_ = mask; }
+    const std::vector<glm::uint8>& getMask() const { return binaryMask_; }
+
+    void setTensors(std::shared_ptr<std::vector<matN>> tensors);
+    void setMetaData(std::shared_ptr<DataFrame> metaData);
+
+    /*
+    If the tensor field has a mask, this method return the number of 1s in it -
+    telling you how many of the positions in the tensor field are defined.
+    */
+    int getNumDefinedEntries() const;
+
+    /**
+     * Data map for the eigen values of the tensor field.
+     * 0 := major, 1 := middle, 2 := minor
+     */
+    std::array<DataMapper, N> dataMapEigenValues_;
+    /**
+     * Data map for the eigen vectors of the tensor field (global min/max of all vector components).
+     * 0 := major, 1 := middle, 2 := minor
+     */
+    std::array<DataMapper, N> dataMapEigenVectors_;
+
+    bool hasMask() const { return binaryMask_.size() == size_; }
+
+    const util::IndexMapper<N>& indexMapper() const { return indexMapper_; }
+
+    /**
+     * Perform lookup as to whether the specified meta data is available for the tensor field.
+     * HINT: If it is not, you might want to add a meta data processor to your network to calculate
+     * the desired meta data.
+     */
+    template <typename T>
+    bool hasMetaData() const;
+
+    bool hasMetaData(const std::string& name) const;
+
+    /**
+     * Tensor field meta data is stored in a DataFrame. If avaliable, this method returns the column
+     * for the meta data specified by T (see attributes.h). Nullopt otherwise.
+     */
+    template <typename T>
+    std::optional<std::shared_ptr<const Column>> getMetaData() const;
+
+    /**
+     * This method returns the underlying container for the meta data specified by T (see
+     * attributes.h). NOTE: This method does not check whether or not the specified meta data is
+     * available.
+     */
+    template <typename T>
+    const std::vector<typename T::value_type>& getMetaDataContainer() const;
+
+    std::shared_ptr<const DataFrame> metaData() const { return metaData_; }
+    std::shared_ptr<DataFrame> metaData() { return metaData_; }
+
+    value_type getMajorEigenValue(const size_t index) const {
+        return value_type(this->getMetaDataContainer<attributes::MajorEigenValue>()[index]);
+    }
+    value_type getMinorEigenValue(const size_t index) const {
+        return value_type(this->getMetaDataContainer<attributes::MinorEigenValue>()[index]);
+    }
+
+    const std::vector<vecN>& majorEigenVectors() const {
+        return this->getMetaDataContainer<attributes::MajorEigenVector<N>>();
```

Comment:
what happens if the metadata does not have eigen vectors?

Reply:
Tensor fields always have eigen values and eigen vectors as these are computed on construction.